### PR TITLE
misc(v2): add more typing for server code

### PR DESCRIPTION
**Pull request by yangshun**
## Motivation

Better typing ftw! With this PR, the server-side code should be mostly covered.

So that we can move fast and don't break things 🔥 

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/Docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

Site builds. No errors.

## Related PRs

NA

### CODE REVIEW

File: packages/docusaurus/src/server/plugins/index.ts [hunk 1]
```diff
@@ -5,21 +5,51 @@
...
-import {LoadContext} from '..';
+
+export interface Plugin<T> {
+  name: string;
+  loadContent?(): T;
+  contentLoaded?({
+    content: T,
+    actions: DocusaurusPluginContentLoadedActions,
+  }): void;
+  postBuild?(props: any): void;
```
Comment (yangshun) [inline packages/docusaurus/src/server/plugins/index.ts]: Is this type `Props`? Not very familiar with this part.

Comment (endiliey) [reply packages/docusaurus/src/server/plugins/index.ts]: i think so.

Comment (yangshun) [reply packages/docusaurus/src/server/plugins/index.ts]: Ok added more types. I'm glad we have less `any` now.

File: packages/docusaurus/src/server/plugins/index.ts [hunk 2]
```diff
@@ -5,21 +5,51 @@
...
+export interface Plugin<T> {
+  name: string;
+  loadContent?(): T;
+  contentLoaded?({
+    content: T,
+    actions: DocusaurusPluginContentLoadedActions,
+  }): void;
+  postBuild?(props: Props): void;
+  postStart?(props: Props): void;
+  configureWebpack?(config: DocusaurusConfig, isServer: boolean);
```
Comment (endiliey) [inline packages/docusaurus/src/server/plugins/index.ts]: not docusaurus config but actual  webpack.configuration. result is also webpack config

**Pull request merged**